### PR TITLE
Support split test run

### PR DESCRIPTION
**Pull request by KAUTH**
### Description

Using `pytest-split` we can make the test duration of Functional tests that run on CI smaller, thus improving the development experience.

For Functional tests that we run from `repository-service-tuf-api`:
~6 min total duration for all Functional Tests to complete ([link](https://github.com/repository-service-tuf/repository-service-tuf-api/actions/runs/7602466421)) vs ~11-15 min (there's quite some deviation)

#### Improvements
There's still room for improvement by saving the test duration files, so that `pytest-split` can split the tests in better groups.
We can also flag the slow tests.

#### Testing it out & Performance
[Umbrella](https://github.com/repository-service-tuf/repository-service-tuf):
Sample run-times:
  - Before: [9m 51s](https://github.com/repository-service-tuf/repository-service-tuf/actions/runs/7159210504/usage)
  - After: [5m 15s](https://github.com/repository-service-tuf/repository-service-tuf/actions/runs/7603429978/usage)

Before merging we should already have prepared the CI for all the other services under the umbrella:
- [ ] API: https://github.com/repository-service-tuf/repository-service-tuf-api/pull/519 ([workflow run](https://github.com/repository-service-tuf/repository-service-tuf-api/actions/runs/7603779717))
Sample run-times:
  - Before: [10m 31s](https://github.com/repository-service-tuf/repository-service-tuf-api/actions/runs/7117579375/usage)
  - After: [5m 59s](https://github.com/repository-service-tuf/repository-service-tuf-api/actions/runs/7603779717/usage)
- [ ] Worker: https://github.com/repository-service-tuf/repository-service-tuf-worker/pull/440 - I don't have access to trigger and test it (see [here](https://github.com/repository-service-tuf/repository-service-tuf-worker/actions/workflows/review-approved.yml))
Sample run-times:
  - Before: [19m 46s](https://github.com/repository-service-tuf/repository-service-tuf-worker/actions/runs/7260720862/usage)
  - After: N/A
- [ ] CLI: https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/492 ([workflow run](https://github.com/repository-service-tuf/repository-service-tuf-cli/actions/runs/7603487240))
Sample run-times:
  - Before: [9m 36s](https://github.com/repository-service-tuf/repository-service-tuf-cli/actions/runs/7263272330/usage)
  - After: [5m 4s](https://github.com/repository-service-tuf/repository-service-tuf-cli/actions/runs/7603487240/usage)

_Note for review: Check that in the end all tests are run after being split_

### Diagram

```mermaid
flowchart TB
    %% A sub-repo such as RSTUF-CLI
    subgraph "RSTUF sub-repo CI (e.g., CLI)"
    
    functional-tests

    %% A sub-repo such as RSTUF-CLI
    functional-tests-slow
    end

    %% Slow tests are ran separately
    functional-tests-slow -->|Run| slow-ci-job[Slow CI job]
    
    functional-tests -->|Checkout RSTUF umbrella| ci-job-group-one[CI job]
    functional-tests -->|Run|ci-job-group-two[CI job]
    functional-tests -->|Run| ci-job-group-three[CI job]

    ci-job-group-one -->|Run with latest RSTUF services| docker-services[Docker services]
    ci-job-group-one -->|SLOW: bool, GROUP: int - parameters passed| pytest[Run tests]

    rstuf-umbrella(Checked out RSTUF umbrella) -.->|Functional tests| pytest
    pytest -.-> docker-services
```

### CODE REVIEW

**Comment by MVrachev** [comment]
When should we start reviewing this and it's related prs?
I guess when you fix the CI issues everywhere?

**Comment by KAUTH** [comment]
> When should we start reviewing this and its related prs? I guess when you fix the CI issues everywhere to make sure these are final versions ready to go?

The only one I see failing now is the `Lint` in this PR, but locally it passes for me.... Maybe we need to merge this https://github.com/repository-service-tuf/repository-service-tuf/pull/596 and rebase?

File: .github/workflows/functional.yml [hunk 1]
```diff
@@ -70,11 +73,60 @@ jobs:
       - name: Bootstrap/Setup RSTUF using DAS and run Functional Tests
         timeout-minutes: 30
         run: |
-          docker compose run --env UMBRELLA_PATH=. --rm rstuf-ft-runner bash tests/functional/scripts/run-ft-das.sh ${{ inputs.cli_version }}
+          docker compose run --env UMBRELLA_PATH=. --rm rstuf-ft-runner bash tests/functional/scripts/run-ft-das.sh ${{ inputs.cli_version }} ${{ matrix.pytest-group }}
```
Comment (MVrachev) [inline .github/workflows/functional.yml:76]: What will happen when you pass the `${{ matrix.pytest-group }}` to docker compose?

File: .github/workflows/functional.yml [hunk 2]
```diff
@@ -53,6 +53,9 @@ jobs:
   functional-das:
     name: "DAS"
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        pytest-group: [ "1", "2", "3"]
```
Comment (MVrachev) [inline .github/workflows/functional.yml:58]: I have a few questions here:
1. how did you decide to use three groups
2. why do you need names for those names and how are those names used?

File: Makefile [hunk 3]
```diff
@@ -56,24 +56,29 @@ lint:
 	rm requirements.commit
 
 functional-tests-exitfirst:
-	pytest --exitfirst --gherkin-terminal-reporter tests -vvv --cucumberjson=test-report.json --durations=0 --html=test-report.html
+ifeq ($(SLOW),)
+	echo "Running fast tests"
+	pytest --exitfirst --splits 3 --group $(PYTEST_GROUP) --store-durations --durations-path=./.test_durations.$(PYTEST_GROUP) --deselect=tests/functional/targets/test_performance.py::test_api_requester_multiple_request_and_targets[50-50-600] tests -vvv --cucumberjson=test-report.json --durations=0 --html=test-report.html
```
Comment (MVrachev) [inline Makefile:61]: What's the benefit of providing the `PYTEST_GROUP` argument?

File: Makefile [hunk 4]
```diff
@@ -56,24 +56,29 @@ lint:
...
 functional-tests:
 	pytest --gherkin-terminal-reporter tests -vvv --cucumberjson=test-report.json --durations=0 --html=test-report.html
 
-
 # CLI_VERSION enables using specific RSTUF CLI version, default: dev
 # usage: `make ft-das CLI_VERSION=v0.8.0b1`
 ft-das:
 ifneq ($(CLI_VERSION),)
-	docker compose run --env UMBRELLA_PATH=. --rm rstuf-ft-runner bash tests/functional/scripts/run-ft-das.sh dev
+	docker compose run --env UMBRELLA_PATH=. --rm rstuf-ft-runner bash tests/functional/scripts/run-ft-das.sh dev $(PYTEST_GROUP) $(SLOW)
```
Comment (MVrachev) [inline Makefile:74]: How does docker-compose use `PYTEST_GROUP` and `SLOW`?

**Review by MVrachev — COMMENTED** [review]
I need introduction to these things and that's why I asked a lot of questions.

File: .github/workflows/functional.yml (hunk 1 above)
Comment (KAUTH) [reply .github/workflows/functional.yml:76]: It's actually being passed to `tests/functional/scripts/run-ft-das.sh` (as we did also with `inputs.cli_version`) and from that bash script we call the `Makefile` along with those parameters ([here](https://github.com/repository-service-tuf/repository-service-tuf/blob/7f004af8bb3026f505ef464e191f1524add2c221/tests/functional/scripts/run-ft-das.sh#L93)).
In the end, what's being called is: `pytest --exitfirst --splits 3 --group $(PYTEST_GROUP) ...` and the `${{ matrix.pytest-group }}` you asked about is the parameter that defines the testing group that is being run at that moment. `pytest-split` splits the tests in groups and on every pytest run you define which group you run.

File: .github/workflows/functional.yml (hunk 2 above)
Comment (KAUTH) [reply .github/workflows/functional.yml:58]: 1. 3 groups was a good balance considering the shortest time the tests take to run and the total time (sum of all jobs), but we can try other number of groups as well.
2. For `pytest-group: [ "1", "2", "3"]`: "1", "2", "3" are not used as names, but they are passed via `${{ matrix.pytest-group }}` to the test script as numbers and are used to define which testing group is running on every job. See https://jerry-git.github.io/pytest-split/#usage.

File: Makefile (hunk 3 above)
Comment (KAUTH) [reply Makefile:61]: It's necessary when using the `pytest-split` plugin:

So if you choose to split your tests into 3 groups, you will have something like:
```shell
pytest --splits 3 --group 1
pytest --splits 3 --group 2
pytest --splits 3 --group 3
```

See https://jerry-git.github.io/pytest-split/#usage

File: Makefile (hunk 4 above)
Comment (KAUTH) [reply Makefile:74]: docker-compose is starting the [`rstuf-ft-runner` service ](https://github.com/repository-service-tuf/repository-service-tuf/blob/a5f92f092ec22d8e4cc8b6dda1925ca22755357b/docker-compose.yml#L76) and runs the command `bash tests/functional/scripts/run-ft-das.sh dev $(PYTEST_GROUP) $(SLOW)`.
`PYTEST_GROUP` and `SLOW` are passed as parameters to the `tests/functional/scripts/run-ft-das.sh` bash script and are later on used to:
- define the testing group for `pytest-split` 
- run the slowest by far test in a separate job

respectively.

**Comment by KAUTH** [comment]
> I need introduction to these things and that's why I asked a lot of questions.

Sure, ask me anything that's not clear 👍

File: Makefile (hunk 3 above)
Comment (ivanayov) [reply Makefile:61]: Isn't it better to use `least_duration` instead of an absolute order? It's more effective over time

**Comment by ivanayov** [comment]
Bringing it here, as we discussed, so it's not lost in threads.

Is there a strong reason behind using an absolute order? 
Isn't it better to use `least_duration` instead, as it's more effective over time and would require less support (like changing order manually to improve performance)?

File: Makefile (hunk 3 above)
Comment (MVrachev) [inline Makefile:61]: Could you please leave a comment where this `--splits 3` comes from?

**Review by MVrachev — COMMENTED** [review]
Could you add some documentation about the tests and how we run them:
https://repository-service-tuf.readthedocs.io/en/stable/devel/development.html#functional-tests-using-behavior-development-driven-bdd
Also, how to include a new slow tests and what is the best practises with pytest-split?
It will be good to mention that if we add a new slow test with new group or job, then we would need to change all YML files in all repositories.

**Comment by KAUTH** [comment]
> Could you add some documentation about the tests and how we run them:
https://repository-service-tuf.readthedocs.io/en/stable/devel/development.html#functional-tests-using-behavior-development-driven-bdd

Added the relevant documentation

> Also, how to include a new slow tests and what is the best practises with pytest-split?
It will be good to mention that if we add a new slow test with new group or job, then we would need to change all YML files in all repositories.

I would include more details as a follow-up task after this is merged in all repos. It would be a good opportunity to explain better how our test suite is structured.

**Comment by KAUTH** [comment]
With the changes in the `Makefile` running locally might need to run `make` manually with different arguments in order to run all test sub-groups. We can improve this later on to be able to choose to run everything at once.

**Review by kairoaraujo — APPROVED** [review]

**Pull request merged**